Announcing the FFI-unwinding Project Group (#2797)
rust-lang/rfcs: text/0000-project-ffi-unwind.md

- Feature Name: `project-unwind-FFI`
- Start Date: 2019-10-27
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

* To create a "project group" with the purpose of designing subsequent RFCs to

>   flesh out the details of the "C unwind" ABI

Reviewer: ```suggestion
  flesh out the details of the "C unwind" ABI or equivalent language feature
```

Author: Ah, good catch. Better yet, here's the top-level summary of our charter:

```suggestion
  extend the language to support unwinding that crosses FFI boundaries
```

  * The "project group" term is newly introduced: it is a specific type of
    working group whose goal is to flesh out a particular proposal or complete
    a project.

>   * This project group plans to specify how "C unwind" works on major

Reviewer 2: ```suggestion
  * This project group plans to recommend specifications of how "C unwind" works on major
```

>   * This project group plans to recommend specifications of how "C unwind" works on major

Author: ```suggestion
  * This project group plans to recommend specifications of how "C unwind" will work on major
```

caller. Unfortunately, previous attempts to stabilize this behavior have caused
existing, working projects to break.

The problem here is not that the existing projects break *per se*: they are
relying on [Undefined Behavior], so breakage is to be expected as a
possibility. The problem is that there is no alternative available to them that
would allow them to keep working (even if they are continuing to rely on
behavior that is not yet fully specified).

Previous attempts to provide a well-defined mechanism for unwinding across FFI
boundaries have failed to reach consensus. Notably, two proposed RFCs generated
over 400 comments between them before ultimately being closed:

* [rust-lang/rfcs#2699](https://github.com/rust-lang/rfcs/pull/2699)
* [rust-lang/rfcs#2753](https://github.com/rust-lang/rfcs/pull/2753)

GitHub comment threads become difficult to follow for discussions this lengthy,
and the disagreements in these threads have felt less productive than we
believe they could be if more structure were provided.

We would also like to demonstrate the Rust lang team's commitment to providing
such a mechanism without needing to agree in advance on what language changes

> will be stabilized in order to do so.

Reviewer 2: ```suggestion
will be stabilized in order to do so. Note however that the lang team's
commitment to some solution is not a commitment to *any* solution.
```
This should be obvious but I'd like to clarify this in the text nonetheless.

Author: I'm actually confused by this sentence; what do you mean by "any solution"? Are you just saying "we won't rush to stabilize a solution"?

Reviewer 2: It just means that while we'll strive to find some acceptable solution, there may be solutions that are not acceptable.

Reviewer 2: But I guess this can be left implicitly understood. :)

## What is a "project group"?

The "project group" term has not previously been used: it is intended to
formalize a concept that has existed informally for some time, under a number
of names (including "working group").

A "project group" is a group of people working on a particular project at the
behest of an official Rust team. Project groups must have:

* A **charter** defining the project's scope
* A **liaison** with an official Rust team (who may or may not also be a shepherd)
* A small number of **shepherds**, who are responsible for summarizing
  conversations and keeping the lang team abreast of interesting developments.
* A GitHub repository hosted under the `rust-lang` organization containing the
  charter and instructions for how community members can monitor the group's
  progress and/or participate.

[This blog post][shepherds-3.0] explains in detail the role of the
shepherds.

## Project group roadmap and RFCs

The first step of the project group is to define a **roadmap** indicating the
planned sequence in which it will design and propose particular behaviors and
features.  Once the project group feels it has completed work on some item in

> the roadmap, that item will be submitted to the community as an RFC or FCP.

Reviewer 2: ```suggestion
the roadmap, that item will be submitted to the community as an RFC or FCP,
and to be reviewed by the lang team.
```

Author: I was considering the lang team to be part of "the community" here. Does this need to be called out explicitly, since, ultimately, the lang team needs to weigh in on such RFCs anyway? (Is that correct?)

Reviewer 2: I only wanted to add this because the current wording suggests that the PG would be the ones approving of the FCP. Here's an alternate wording:
```suggestion
the roadmap, that item will be submitted as an RFC or FCP for review by the lang team and the community at large.
```

## Stabilizing unspecified "TBD" behavior
[stabilizing-tbd]: stabilizing-unspecified-tbd-behavior

We would like to be able to provide features in stable Rust without a full
formal specification but with an informal statement of intent regarding the
behavior. These features would be considered "unspecified behavior" (rather

> than "undefined behavior"), and their behavior would therefore be subject to

Reviewer 2: As worded right now, this sort of assumes that everyone agrees what "unspecified" means (that often turns out not to be the case). I assume this will be detailed and clarified as the project progresses. I hope this will not turn into "implementation defined" or some such.

Author: We have indeed written up [our own definition](https://github.com/rust-lang/project-ffi-unwind/blob/master/spec-terminology.md), and I think it would be worthwhile to open an RFC to concretely define this terminology.

I think "implementation defined" would be a good addition to the `spec-terminology` doc.

Reviewer: An alternative is for "TBD" to be an orthogonal denotation to whatever the behavior class used by the reference is (unspecified, undefined, target-dependent, ...). 

One of the consequences of unspecified meaning different things to different people (and different WGs, the UCGs has its own definitions and they are controversial!), is that users might end up confused on whether "unspecified" behavior can become "undefined" in the future or not.

I don't think that's something that's worth resolving right now, and I think that it is in the best interest of the Project Group to have as much freedom as possible. By making it an orthogonal annotation, the Project Group can say that "X is UB (TBD) and Y is target-dependent (TBD)" meaning that X is UB and Y is target dependent, and that the project group is trying to improve both, but while X might remain UB, Y won't become worse than target-dependent.

Author: Hm. I appreciate the suggestion, @gnzlbg, but I don't really think that helps with the problem Niko and I were trying to solve by defining/adopting "TBD", which is to have a way of saying, essentially, "this is _not_ UB in the sense that the compiler may optimize based on the assumption that it can't happen; we have not fully specified the behavior, but you can expect it to 'work' approximately as you'd expect, and we promise to further specify the behavior later (though until that time the details of the behavior-as-implemented may change)."

I don't think it would be appropriate to define "unspecified behavior" as part of this "announcement" RFC, but if @Centril (or someone else) would prefer, I can just remove this section entirely (assuming @nikomatsakis agrees that's okay).

Reviewer: That would mean that any behavior that we might want to define in some way, and one of the alternatives being considered is defining it as "UB", would need to be UB instead of TBD. If that's the intent, then that's fine with me. Maybe add a sentence mentioning just that ?

Reviewer 2: @BatmanAoD I'm just happy with the knowledge that you'll take into account that I don't think this should become implementation-defined and that the lang team should have the freedom to change the Rust unwind ABI on any platform including with shims if necessary.

Reviewer 3: My two cents:

I think it is useful to enumerate our scope and goals and to communicate those goals to people. So I think that distinguishing TBD is very useful, as it communicates that. I think that if we can give people guidelines as to our *intent* (along the lines that @BatmanAoD sketched -- perhaps more specific) that's even better.

That said, I think we should leave ourselves the freedom to make decisions. I think we can define our intent here as roughly:

* We intend to make it possible for Rust panics and foreign exceptions to propagate through a "C unwind" boundary in a well-defined and specified way.
* As a non-exhaustive list, this means that we intend to eventually use cases like the following:
    * a C++ exception unwinding Rust frames, running destructors as it goes
    * a Rust panic unwinding through C++ frames, running destructors as it goes
    * a Rust panic unwinding into C++ and then back into Rust, and being caught via `catch_unwind`
* However, there are a number of details whose resolution is not clear. As a non-exhaustive list:
    * what should happen when a C++ exception attempts to unwind through a Rust `catch_unwind` call?
    * how should longjmp, which is defined on some platforms via unwinding, interact with frames that contain destructors? (And is that even in scope for this group?)

Along those lines, I think we can specify that:

* For Tier 1 platforms, there will be *some* translation from Rust panics into the native unwinding mechanism, but we have not yet specified how Rust panics will be described
* Similarly, we have not specified what the behavior will be when unwinding across a "C unwind" barrier occurs, except for the case where the exception was originally a Rust panic, in which case it is translated back into a Rust panic
    * This implies that we have not specified what happens when `catch_unwind` is involved, since before `catch_unwind` gets involved, there must be a "C unwind" barrier.

I'm going to be honest and say that the distinction between "unspecified behavior" and "undefined behavior" feels like a "distinction without a difference" to me, most of the time, unless that unspecified behavior is further qualified to some narrower range of possibilities. I guess I would mostly rather we just don't waste a lot of time (too late...) arguing back and forth on this, as I'm not sure what practical import it has. I'd rather we just spend our time deciding what the behavior *ought to be*.

reference material for the language, the behavioral intent as determined by the
project group should be described, but will not imply any special status
compared to other instances of unspecified behavior in the language.

## Details of the FFI-unwind project group

[Repository][ffi-unwind project]

Initial shepherds:

* [acfoltzer (Adam)](https://github.com/acfoltzer)
* [batmanaod (Kyle)](https://github.com/batmanaod)

Lang team liaisons:

* [nikmoatsakis (Niko)](https://github.com/nikmoatsakis)
* [joshtriplett (Josh)](https://github.com/joshtriplett)

### Charter
[charter]: #charter

The FFI-unwind project group has the following initial scope:

* to define the details of the "C unwind" ABI on major Tier 1 platforms
* in particular, to define with sufficient detail to enable the use cases
  described in the Motivation section of this RFC
  
Certain elements are definitively out of scope:

* The group does not intend to consider mechanisms to enable "interop"
  between Rust panics and exceptions from other languges. For example,
  we do not intend to permit Rust code to catch C++ exceptions, though
  we will have to consider what happens when a C++ exception unwinds

>   past a `catch_unwind` boundary.

Reviewer: There are a couple of contradictions in this section.

First, the Motivation only speaks about "unwinding" out of `extern "C"` functions in general, and as a general problem, that includes dealing with foreign exceptions (e.g. due to cancellation points on Linux). 

Second, this paragraph suggests that foreign exceptions will be ignored while simultaneously saying that they will be considered. 

I think this section should clarify whether the intent is for the project group to provide features that will work with foreign exceptions or not at some point. While it is fair to say that supporting foreign exceptions is not part of an MVP, whether they are factored in the initial design or not might be the difference between `"C unwind"` supporting them in the future, or the language having to add a different language feature to support them (e.g. another `"C platform-unwind" ABI).

Author: > First, the Motivation only speaks about "unwinding" out of `extern "C"` functions in general, and as a general problem, that includes dealing with foreign exceptions (e.g. due to cancellation points on Linux).

I'm not sure that's a "contradiction"; it just means that some related features that could be helpful are nevertheless out of scope for _this_ group. However, given Amanieu's PR, maybe it's not a great example of something that will _never_ be in scope.

Author: > Second, this paragraph suggests that foreign exceptions will be ignored while simultaneously saying that they will be considered.

I don't know why you would say "foreign exceptions will be ignored"; this section is just saying that full exception-interop (e.g. defining how to interact with foreign exception objects) is not in scope.

> I think this section should clarify whether the intent is for the project group to provide features that will work with foreign exceptions or not at some point.

Well, as it stands, it does clarify that, by saying that the group won't provide such features. But if you're saying that this RFC _must_ specify whether or not these features will ever be in scope for the group, then I disagree; I don't see what benefit there would be in saying "that's in scope" or "that's out of scope" before the group has proposed a single technical RFC.

Reviewer: Maybe this section should explain in more detail what precisely is in scope and what precisely is not.

> I don't know why you would say "foreign exceptions will be ignored";

That's what this sentence says to me: "The group does not intend to consider mechanisms to enable "interop" between Rust panics and exceptions from other languge". I read that as this working group will not deal with how exceptions from other languages, called "foreign exceptions", interoperate with Rust. Yet the next sentences say that in some cases, this working group will deal with precisely that. 

If that isn't what this sentence tries to convey, what is it? Maybe it could be worded somehow else to prevent me from misunderstanding it in that way ?

Reviewer 3: I think this is another case where it's useful to define expectations. Most important seems to be saying what we DO plan to do. It's also useful to say what we don't *expect* to do, but we can leave room for future extension. I guess here I would say something like:

We expect to define how foreign exceptions interact with `catch_unwind`. We do not currently expect to add new mechanisms for interacting with or throwing foreign exceptions. We will revisit possible extensions when the initial goals have been met.

### Constraints and considerations

In its work, the project-group should consider various constraints and
considerations:

* The possibility that C++ may adopt new unwinding mechanisms in the future.
* The possibility that Rust may alter its unwinding mechanism in the future --
  in particular, the project group must not propose a design that would

>   constrain Rust's unwinding implementation.

Reviewer 2: ```suggestion
  constrain Rust's unwinding implementation, including on major Tier 1 platforms.
```

Author: How about "...implementation on current or future platforms"? Calling out Tier 1 here seems to imply that perhaps they _should_ be treated differently in some way, which isn't the intent of this section.

Reviewer 2: I used "Tier 1" here because it is specifically called out in the Charter. I'd be fine with "on current or future platforms" but I've clarified previously how I feel about Tier 1 platforms as well on Zulip so y'all can take that into account in the PG.

Author: Sorry, I don't remember comments from you about Tier 1 platforms, and I can't find them with the search function! Care to reiterate?

Reviewer 2: I think we should be free to change Rut's unwinding implementations on Tier 1 platforms as well. Otherwise the freedom suggested here wouldn't mean much in practice.

Reviewer: >  constrain Rust's unwinding implementation, including on major Tier 1 platforms.

@Centril I think the "including on major Tier 1 platforms" is unnecessary, the intent is to avoid constraining the Rust unwind implementation on any platform, not only Tier 1, and that's how I understood the bullet point. If that was unclear, maybe:

```suggestion
  constrain Rust's unwinding implementation on any target.
```

is clearer ?

Author: @Centril I've adopted @gnzlbg's suggested wording.

Reviewer 2: I'm happy with that.

In addition to the [blog post Niko Matsakis][shepherds-3.0] about
shepherding, James Munns wrote a [more formal shepherding
proposal][shepherding-3.1].

The [governance WG][governance-wg] and [lang-team meta working
group][lang-meta-wg] were both formed at least in part to improve the process
for large-scale design efforts. One existing proposal is for ["staged
RFCs"][staged-rfc]; this may be considered a precursor to the current
"shepherded project group" proposal.

# Unresolved questions and Future possibilities
[unresolved-questions]: #unresolved-questions

Since this RFC merely formalizes the creation of the project group, it
intentionally leaves all technical details within the project's scope
unresolved.

# Future possibilities
[future-possibilities]: #future-possibilities

The project group will start with a fairly [limited scope][charter], but if the
initial effort to design and stabilize a safe cross-language unwinding feature
on a limited set of platforms goes well, there are many related areas of
potential exploration. Three noteworthy examples are:

* Catching foreign unwinding (e.g. Rust catching C++ exceptions, or C++
  catching Rust `panic`s)

> * Defining subtype relationships among `fn`s using ABIs with different `unwind`

Reviewer 2: Is this about *top level* coercions? I don't think anyone has proposed that e.g. `x: Vec<extern "C" fn(...)> as Vec<extern "C unwind" fn(...)>` be a legal transformation.

Author: I don't know what "top level" means here, and I don't actually know the full rules for covariance in Rust (e.g. would making `extern "C" fn as extern "C unwind" fn` automatically imply that your example would also be valid?).

As far as I can remember, previous discussions on ABI-to-ABI coercions involving `unwind` have mostly been limited to agreeing that for simplicity's sake we shouldn't even worry about them yet, since it's always possible to wrap a function pointer in a lambda with a different ABI. I don't think any additional details or constraints are warranted in this document.

Reviewer 2: ```rust
fn main() {
    let f: fn() = main;
    let uf: unsafe fn() = f; // OK.
    //~^ Top level coercion.

    let vf: &fn() = &f;
    let vuf: &unsafe fn() = vf; // ERROR
    //~^ Subtyping. A reference `&'a T` is covariant wrt. `T`.
    // That means that if `X <: Y` we can take a value `x: &'a X`
    // and use it where `&'a Y` is expected.
    // However, no subtyping relationship holds between `fn() <: unsafe fn()`.
    // Therefore, `&'a fn()` may not be used where `&'a unsafe fn` is wanted.
}
```

> I don't think any additional details or constraints are warranted in this document.

Sure, but I think it would be more realistic if the document said:

```suggestion
* Defining coercions among `fn`s using ABIs with different `unwind`
```

Reviewer: > Is this about top level coercions? I don't think anyone has proposed that e.g. x: Vec<extern "C" fn(...)> as Vec<extern "C unwind" fn(...)> be a legal transformation.

@Centril somebody did: https://github.com/rust-lang/project-ffi-unwind/pull/9/files#diff-1b91cd2b717d4c4e383aec41a4ad206eR304

Author: @Centril Thanks for the clarification. I didn't realize there were cases where implicit coercion could be used without a subtyping relationship. I can't say I fully understand that yet, but I've changed the wording appropriately.
